cpp/include/cuspatial/detail/algorithm/is_point_in_polygon.cuh

```diff
@@ -99,5 +100,96 @@ __device__ inline bool is_point_in_polygon(vec_2d<T> const& test_point, PolygonR
   return point_is_within;
 }
 
+/** @brief check if p3 is on the left side of the arc that is
+ * defined but p1 and p2
+ * @note: T can be float or double and T4 can be float4 or double4
+ * @param p1: first point of an arc
+ * @param p2: second point af an arc
+ * @param p3: a point to check
+ * @return bool
+ */
+template <typename T>
+__device__ bool is_left(vec_3d<T> const p1, vec_3d<T> const& p2, vec_3d<T> const& p3)
+{
+  vec_3d<T> po  = {0, 0, 0};
+  auto p1o      = po - p1;
+  auto p1p2     = p2 - p1;
+  auto p1p3     = p3 - p1;
+  auto p1oxp1p2 = cross_product(p1o, p1p2);
+  auto w        = dot(p1oxp1p2, p1p3);
+
+  return w > 0;
+}
+
+/** @brief check if two arcs, defined with (p1, p2) and (p3, p4)
+ * intersects (true or not)
+ * @note: T can be float or double and T4 can be float4 or double4
+ * @param p1: first point of first arc
+ * @param p2: second point af first arc
+ * @param p3: first point of second arc
+ * @param p4: second point af second arc
+ * @return bool
+ */
+template <typename T>
+__device__ bool is_intersecting(vec_3d<T> const& p1,
+                                vec_3d<T> const& p2,
+                                vec_3d<T> const& p3,
+                                vec_3d<T> const& p4)
+{
+  auto p1left = is_left(p3, p4, p1);
+  auto p2left = is_left(p3, p4, p2);
+  auto p3left = is_left(p1, p2, p3);
+  auto p4left = is_left(p1, p2, p4);
+
+  auto alpha = dot(p1, p3);
+
+  return (p1left != p2left) && (p3left != p4left) && alpha > 0;
+}
+
+/**
+ * @brief Test if a point is inside a polygon on a spherical geometry.
+ *
+ * Implements another "Crossings test" algorithm by picking the first point
+ * of the query polygon as the reference point
+ *
+ * @tparam T type of coordinate
+ * @tparam PolygonRef polygon_ref type
+ * @param test_point point to test for point in polygon
+ * @param polygon polygon to test for point in polygon
+ * @return boolean to indicate if point is inside the polygon.
+ * `false` if point is on the edge of the polygon.
```

Comment:
Document the behavior, not the return type. Doxygen does that for us.

```suggestion
 * @return True if point is inside the polygon, false if point is outside or on the edge.
```

Reply:
Are you absolutely sure this test consistently returns false if the point is on the polygon boundary? This is important to know for DE-9IM. We have had trouble with this with 2D PiP.

Reply:
@ayasar70 can you respond to this one?

Reply:
Yes. Whenever a point is on an arc of the query polygon, ``is_left`` function should return false since the angle between point and orthogonal vector is going to be 90 degrees. Since we only use multiplication and addition, we should be ok in terms of precision. I am going to double check to see if precision can cause a problem.